src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,168 @@
+            this.elementData = grow(s + addLength);
+        }
+        System.arraycopy(a, 0, this.elementData, s, addLength);
+        size = s + addLength;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T get(int index) {
-        return null;
+        rangeCheckForGet(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheckForGet(index);
+        elementData[index] = value;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheckForGet(index);
+        T oldValue = (T) elementData[index];
+        fastRemove(index);
+        return oldValue;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(T element) {
-        return null;
+        final int index = findElement(element);
+        T oldValue = (T) elementData[index];
+        fastRemove(index);
+        return oldValue;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        return Arrays.toString(elementData);
+    }
+
+    @Override
+    public Object[] toArray() {
+        return Arrays.copyOf(elementData, size);
+    }
+
+    private Object[] grow(int minCapacity) {
+        return elementData = Arrays.copyOf(elementData, newCapacity(minCapacity));
+    }
+
+    private int newCapacity(int minCapacity) {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity - minCapacity <= 0) {
+            if (Arrays.equals(elementData, DEFAULTCAPACITY_EMPTY_ELEMENTDATA)) {
+                return Math.max(DEFAULT_CAPACITY, minCapacity);
+            }
+            if (minCapacity < 0) {
+                throw new OutOfMemoryError();
+            }
+            return minCapacity;
+        }
+        return (newCapacity - Integer.MAX_VALUE <= 0)
+                ? newCapacity : Integer.MAX_VALUE;
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add inappropriate index: "
+                    + index);
+        }
+    }
+
+    private void rangeCheckForGet(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't get inappropriate index: "
+                    + index);
+        }
+    }
+
+    private void fastRemove(int index) {
+        final int newSize = size - 1;
+        if (newSize > index) {
+            System.arraycopy(elementData, index + 1,
+                    elementData, index, newSize - index);
+        }
+        elementData[size = newSize] = null;
+    }
+
+    private int findElement(T o) {
```

Comment:
The `findElement` method is not using informative names for variables and methods. The variable `o` should be renamed to something more meaningful that reflects its purpose, like `elementToFind`.